💄 Ender3v2 DWIN MarlinUI Landscape Improvements

The-EG opened the pull request:
### Description

Some improvements to the MarlinUI implementation Ender 3 v2's DWIN in **landscape** orientation.

Sometimes you have to admit simple is better. I tried to be creative with the original layout, but it was causing issues (fan was overlapping remaining time/show_e_total), so I've rearranged it to be more similar to the portrait and DOGM/12864 layout:
![image](https://user-images.githubusercontent.com/7342077/147508360-772b0649-ca3d-46ff-9a77-67960de80144.png)

With show_e_total and show remaining time, it will look like this during a print:
![image](https://user-images.githubusercontent.com/7342077/147508386-633b893e-73c0-46fd-8cee-62afb2d41cce.png)

I've additionally triggered the 'first draw' again anytime the print job timer starts or stops. This should eliminate any of the issues with the status screen being in an odd/inconsistent state at the end of a print job.

### Requirements

The DWIN with an encoder that comes stock with an Ender 3 v2

### Benefits

The landscape layout is more standard compared to other MarlinUI layouts and:
 - The fan icon/value no longer overlaps the value for remaining time/e total
 - The fan icon will now stop spinning properly when the fan is turned off

### Configurations

Use your favorite E3v2 config with `DWIN_MARLINUI_LANDSCAPE`

### Related Issues

None that I know of.

The-EG committed:
💄 Ender3v2 DWIN MarlinUI Landscape Improvements

thinkyhead commented:
Looks pretty nice. I might want to space out the X Y Z positions a little more. How does the layout look with two or three extruders, and with / without the heated bed?

On a side-note, I have been meaning to add an indicator to show that the SD card is inserted / absent, and an indicator that bed leveling is enabled, by drawing the bed icon differently, perhaps with a different color, or with lines dividing the bed into 4 quadrants, or with colored triangles in the empty spaces left & right of the bed.

The-EG commented:
> Looks pretty nice. I might want to space out the X Y Z positions a little more. How does the layout look with two or three extruders, and with / without the heated bed?

Thanks. It still supports 2 extruders as it did before: the bed icon will be shifted further right to accomodate. I could probably do 3 and still have room for the bed, but the portrait mode definitely wouldn't be able to. In general the heaters are drawn conditionally, so without a heated bed, that whole column is skipped.

> On a side-note, I have been meaning to add an indicator to show that the SD card is inserted / absent, and an indicator that bed leveling is enabled, by drawing the bed icon differently, perhaps with a different color, or with lines dividing the bed into 4 quadrants, or with colored triangles in the empty spaces left & right of the bed.

There's definitely some space to work with here. I remember you saying that about bed leveling before. It'd be an opportune time to add that, an sd card indicator and anything else useful in. 

Are there existing configuration defines for those? Would they be optional or just 'automagic,' it would happen automatically if the appropriate feature is enabled?

thinkyhead commented on Marlin/src/libs/stopwatch.cpp:
```diff
@@ -38,6 +42,7 @@ bool Stopwatch::stop() {
 
   if (isRunning() || isPaused()) {
     TERN_(EXTENSIBLE_UI, ExtUI::onPrintTimerStopped());
+    TERN_(IS_DWIN_MARLINUI, ui.did_first_redraw = false);
```
The `did_first_redraw` flag is meant to be internal and private, not something that is meant for high-level code to hack. If the screen needs to redraw differently due to a change in paused / active printing state, the screen implementation ought to handle this internally. There are no other MarlinUI-based controllers that need or implement a special handler for `onPrintTimerStarted`/`onPrintTimerPaused`/`onPrintTimerStopped` but that's the method that perhaps should be added, with an empty implementation for all screens except this one.

The-EG replied:
OK. My initial thought was to track the state internally every time the status screen was drawn and update `did_first_redraw` as needed, but this occurred to me as a way of not having that check on every draw, but I can see wanting to keep that internal too.

I'll switch it back to that initial approach.

thinkyhead commented:
> Are there existing configuration defines for those?

I don't think they need to be configurable, but should just be part of the standard display behavior when SD card / leveling are on. I don't think anything needs to be added to the `MarlinUI` class to support it. But, for the E3V2 screens, the `DWIN_SET` will need some updates: An extra pair of bed icons for "leveling on", an icon of an SD card, and maybe an icon of a Flash Drive for future use.

Someone had reported that fonts were acting strangely with E3V2 CrealityUI, so it might be a good time to check the integrity of the `DWIN_SET` generally, too.

thinkyhead committed:
cleanup

tome9111991 commented:
> > Are there existing configuration defines for those?
> 
> I don't think they need to be configurable, but should just be part of the standard display behavior when SD card / leveling are on. I don't think anything needs to be added to the `MarlinUI` class to support it. But, for the E3V2 screens, the `DWIN_SET` will need some updates: An extra pair of bed icons for "leveling on", an icon of an SD card, and maybe an icon of a Flash Drive for future use.
> 
> Someone had reported that fonts were acting strangely with E3V2 CrealityUI, so it might be a good time to check the integrity of the `DWIN_SET` generally, too.

Btw
I got issue with font in ENHANCEDUI
https://github.com/mriscoc/Marlin_Ender3v2/issues/180

The-EG commented:
> Btw
> I got issue with font in ENHANCEDUI

The issue with the stock font is that it is missing the special characters that Marlin UI uses for degree, federate, and menu up/select. So while the stock font might work for the stock UI and derivatives, we need the Marlin font for Marlin UI.

The comment regarding summer fonts being wider than others is just plan false, unfortunately. The dwin displays all characters at a fixed width. The content of individual glyphs might be narrower but that just means there's an appearance of more space between glyphs.

In the issue pictured, I suspect the menu item jpg is overlapping the left of the first font glyph a bit. With the stock font it may not be an issue but it's likely still getting covered a bit (it's just the content of the glyph is narrower)

The-EG commented:
An example of what I mean by 'content of the glyph.' In this case the '8' character. Both the stock font and MarlinUI have font glyphs for 12x24 characters, among others. As the size name suggest, each glyph is 12 pixels wide by 24 pixels tall.

The stock font:
![image](https://user-images.githubusercontent.com/7342077/147605250-6bfbbebb-2f8e-40a5-a243-e0b4284ede73.png)

The MarlinUI font:
![image](https://user-images.githubusercontent.com/7342077/147605319-01b8c5ce-4773-482b-b4c1-6e647fdd9f5b.png)

Notice the stock font is shifted a couple pixels to the right. They are both the exact same width, but the stock font looks thinner due to the style and isn't being overlapped because it's got a few pixels to spare on the left hand side.

If these font issues are all involving overlaps like that, I'd suggest fixing the positioning rather than trying to kludge individual font glyphs to work.

The-EG commented:
Whipped up a couple of new icons.

Bed leveling enabled - bed heater off:
![017-ICON_bed_off_leveling](https://user-images.githubusercontent.com/7342077/147611174-296f6a18-6724-43cd-b8b4-f4ea7202bf9f.jpg)
![image](https://user-images.githubusercontent.com/7342077/147611168-c13f9beb-7d20-4281-9c88-82bd86449f4b.png)

Bed leveling enabled -bed heater on:
![017-ICON_bed_on_leveling](https://user-images.githubusercontent.com/7342077/147611196-1eadc030-a783-4568-846a-7267e0e210ce.jpg)
![image](https://user-images.githubusercontent.com/7342077/147611202-fa929163-8663-4bca-ba5e-19533ee8c394.png)

Also, I'm guessing the hotend on fill and bed heater on line stroke color changed at some point since I last worked with these. It appears to be #fee026ff .

thinkyhead commented:
Nice, that's pretty much exactly the bed icon I had in mind.

Yes, I did mess with those colors a bit. Whatever your eyedropper tool says is likely correct, but I'll check in Inkscape and see what the exact value is.

thinkyhead commented:
Here are my SVG files in their current state… [icon-svg.tar.gz](https://github.com/MarlinFirmware/Marlin/files/7787182/icon-svg.tar.gz)

The-EG commented:
> Here are my SVG files in their current state… icon-svg.tar.gz

That still doesn't match what's actually in the DWIN_SET in the config repo :laughing: . The bed and hotend off aren't filled any there. 

Oh well, I think I got close enough in any case. I'll put the 'current' set of SVGs in `buildroot` so that we don't have to track them down next time, hopefully.

The-EG committed:
💄 Add DWIN icon source svgs

The-EG committed:
💄 Add new icons for bed leveling state

The-EG commented:
I've added what should be the source SVGs for the current icons to `buildroot/share/pixmaps/dwin-icons`. This does not include the Marlin icon or the bootscreen graphics. There's also a small script for generating the appropriate JPGs from them using ImageMagick.

The next commit adds in the bed leveling status icons and keeps `ui.did_first_redraw` private by tracking the state of `pring_job_timer.isRunning()` internally.

Also worth nothing: icon 16 (bedline) was missing from the DWIN_SET 4.ICO icons in the config repo. This broke `BABYSTEP_ZPROBE_GFX_OVERLAY` and `MESH_EDIT_GFX_OVERLAY`. Either no one is using those or they haven't noticed...

I've added it back in along with the two new icons. We are getting really close to the 96kB limit (3 'spaces' at 32kB each) for this icon file before we have to start shuffling assets around again (currently 93kB).

[4.ICO.zip](https://github.com/MarlinFirmware/Marlin/files/7789768/4.ICO.zip)

thinkyhead commented:
> That still doesn't match what's actually in the DWIN_SET

I have work scattered around on a few computers, so I'll see if I can find the latest stuff.

thinkyhead committed:
misc. cleanup

thinkyhead commented:
If all is well here, I'll get this merged and bring the `4.ICO` file over to the Configurations repo. I compared some of my DWIN work folders and didn't find major differences, but I have another location to check so might still drop a ZIP here later on.

The-EG commented:
I'm happy where it's at and I think it's an improvement over the current setup. I've even implemented the bed leveling icons for DOGM now so that my ender 3 pro's 12864 display has a similar setup. I didn't realize how much I'd like that feature until I didn't have it! I can PR that too if there's interest.

I'll do some final testing this weekend and verify nothing got jostled during the clean up and merges.

thinkyhead commented:
> I've even implemented the bed leveling icons for DOGM now…

Oh yeah, that would be a great PR. I know many would appreciate the extra info on the screen.

> I'll do some final testing this weekend and verify nothing got jostled during the clean up and merges.

Great. I look forward to merging this for 2.0.9.4 and getting it out to the masses. Cheers!

thinkyhead committed:
Merge 'bugfix-2.0.x' into pr/23369

The-EG commented:
@thinkyhead I had a chance to build this again with the latest merges and all is well. I also tested portrait mode to be sure nothing was inadvertently broken there either.

Please merge when ready.

thinkyhead committed:
tweak dwin stuff

thinkyhead committed:
dwin JPEG images

thinkyhead commented:
Do you have a new `DWIN_SET` already put together which I can use to replace the [`DWIN_SET` in Configurations](https://github.com/MarlinFirmware/Configurations/tree/import-2.0.x/config/examples/Creality/Ender-3%20V2/DWIN_SET)?

thinkyhead commented:
@thisiskeithb — I'd suggest that the "please merge" label should be reserved for things that fix urgent bugs, not just for non-critical PRs that are ready for another - maybe final - review. There is no urgency here, and this particular improvement should not be merged until a new `DWIN_SET` is provided at the Configurations repository.

The-EG commented:
> Do you have a new `DWIN_SET` already put together which I can use to replace the [`DWIN_SET` in Configurations](https://github.com/MarlinFirmware/Configurations/tree/import-2.0.x/config/examples/Creality/Ender-3%20V2/DWIN_SET)?

MarlinFirmware/Configurations#663

thinkyhead committed:
Update README.md
